### PR TITLE
Add SVG support via resvg

### PR DESCRIPTION
**Pull request by RReverser**
This is still missing text rendering because dealing with fonts is its own problem, but otherwise works well and is enough to pass Sharp tests.

Wanted to submit anyway so that my branch / work isn't lost, but leaving disabled by default because it's both incomplete w/o text, but also because it's a relatively rarely useful feature. More common problem is optimising SVG to SVG, for which people use tools like SVGO, while in libvips SVG can only be rasterised and treated like other raster images.

### CODE REVIEW

File: build.sh
Comment (RReverser) [inline build.sh:502]:
```suggestion
    ${ENABLE_SVG:+-Dresvg=enabled} ${ENABLE_AVIF:+-Dheif=enabled} -Dheif-module=enabled
```

**Comment by d3lm** [comment]
Nice, still great to have this! Awesome work 👏

File: Dockerfile
Comment (kleisauke) [inline Dockerfile:10]: `build/cargo-cache` should also be in included in `.gitignore`.

File: build.sh
Comment (kleisauke) [inline build.sh:60]: I think we could enable this by default when building as a dynamic module, allowing users to opt-in by doing:
```js
const vips = await Vips({
  // Enable SVG load support
  dynamicLibraries: ['vips-resvg.wasm']
});
```

Let me know if I should experiment with this. :)

File: build.sh
Comment (kleisauke) [inline build.sh:436]: This comment seems outdated.
```suggestion
```

File: build.sh
Comment (kleisauke) [inline build.sh:437]: Perhaps we should also build with `-Zbuild-std-features=panic_immediate_abort`?
```suggestion
  cargo build --release --target wasm32-unknown-emscripten --locked -Zbuild-std=panic_abort,std -Zbuild-std-features=panic_immediate_abort --no-default-features --features filter,raster-images
```
(untested)

Comment (RReverser) [reply build.sh:436]: It's possible, but the point of that feature is that it swallows the error messages when something does panic. Given that panics occur only on critical bugs, personally I value knowing what happened more than the savings it provides, but I can add it if you think it's worth it.

**Comment by RReverser** [comment]
Just a heads up - might be better to merge https://github.com/kleisauke/libvips/pull/2 before this one to reduce complexity of patches in this PR.

**Comment by kleisauke** [comment]
Great! I was able to support this as a dynamic module and also added some test coverage for this (by porting libvips' SVG tests).
https://github.com/RReverser/wasm-vips/compare/libresvg...kleisauke:wasm-vips:RReverser-libresvg
https://github.com/RReverser/libvips/compare/wasm-vips-resvg...kleisauke:RReverser-wasm-vips-resvg?w=1 (`?w=1` to hide those changes in commit https://github.com/kleisauke/libvips/commit/0b75df254a478b29580ce5d92ba2ed272446883e :sweat_smile:)

I could push these commits to the branches associated with this PR and https://github.com/kleisauke/libvips/pull/2, if that looks good to you.

FWIW, I'm planning to upstream the Rust patches introduced in commit https://github.com/kleisauke/wasm-vips/commit/94e67b1617fc096d2a65b64f4243a7b64f5f4559, here is the relevant CI failure which is the reasoning behind these patches:
https://github.com/kleisauke/wasm-vips/actions/runs/3947805916/jobs/6757635992#step:7:225

File: build.sh
Comment (kleisauke) [reply build.sh:436]: Ah, I see. I don't think we actually need that optimization flag, given that `vips-resvg.wasm` is only 1.4 MB (making it the smallest dynamic module in size so far).
```console
$ ls -lh lib/*.wasm | awk '{print $5,$9}'
5.0M lib/vips-heif.wasm
2.0M lib/vips-jxl.wasm
1.4M lib/vips-resvg.wasm
5.5M lib/vips.wasm
```

File: build.sh
Comment (kleisauke) [reply build.sh:60]: Done.

**Comment by RReverser** [comment]
> here is the relevant CI failure which is the reasoning behind these patches:
> https://github.com/kleisauke/wasm-vips/actions/runs/3947805916/jobs/6757635992#step:7:225

That error looks odd. Why is a Windows CI task using `nightly-x86_64-unknown-linux-gnu` toolchain?

**Comment by RReverser** [comment]
Or, I suppose it's actually Wasm / Emscripten build, but then the tripe should still be Emscripten's I think?

**Comment by kleisauke** [comment]
... the reason that it only fails on Windows is that Unix-based operating systems do have `/dev/urandom` available. I suppose this is not a issue if you do _not_ link with `-sNODERAWFS` or use WasmFS instead.

**Comment by kleisauke** [comment]
With this patch:
```diff
--- a/build.sh
+++ b/build.sh
@@ -171,6 +171,11 @@ fi
 if [ "$WASM_FS" = "true" ]; then export CFLAGS+=" -DWASMFS"; fi
 if [ "$PIC" = "true" ]; then export CFLAGS+=" -fPIC"; fi
 
+# Ensure build paths are removed from the resulting binaries
+# https://reproducible-builds.org/docs/build-path/
+export RUSTFLAGS+=" --remap-path-prefix=$(rustc --print sysroot)/= --remap-path-prefix=$DEPS/="
+export CFLAGS+=" -ffile-prefix-map=$DEPS/="
+
 export CXXFLAGS="$CFLAGS"
 
 export LDFLAGS="$COMMON_FLAGS -L$TARGET/lib -sAUTO_JS_LIBRARIES=0 -sAUTO_NATIVE_LIBRARIES=0"
```

I now see:
```console
$ grep "/usr/local/rustup/toolchains/nightly-x86_64-unknown-linux-gnu" lib/vips-resvg.wasm || echo "no matches"
no matches
```

I'll commit that after this PR lands. :)

**Comment by kleisauke** [comment]
Commit https://github.com/kleisauke/wasm-vips/commit/54bce444a611966e885b053b90f57598b8af2eb1 fixes the remaining tests. I just opened PR https://github.com/libvips/libvips/pull/3307 for that.

**Comment by kleisauke** [comment]
> I'm planning to upstream the Rust patches introduced in commit [94e67b1](https://github.com/kleisauke/wasm-vips/commit/94e67b1617fc096d2a65b64f4243a7b64f5f4559)

I just opened draft PR https://github.com/rust-lang/rust/pull/107221 and https://github.com/rust-lang/libc/pull/3087 for this.

**Comment by RReverser** [comment]
> I just opened draft PR [rust-lang/rust#107221](https://github.com/rust-lang/rust/pull/107221) and [rust-lang/libc#3087](https://github.com/rust-lang/libc/pull/3087) for this.

Nice! Out of curiosity, why do you create those PRs as drafts? They look like should be ready for review straight away?

**Comment by kleisauke** [comment]
> Out of curiosity, why do you create those PRs as drafts? They look like should be ready for review straight away?

The Rust PR depends on the libc PR, so the libc crate needs to be updated first in Rust (i.e. a similar commit as https://github.com/rust-lang/rust/commit/e294640484731dca722f2cc3509a18e8b4a4a20d is needed).

The libc PR was marked as draft because it does not meet this check:
> - \[ ] Edit corresponding file(s) under `libc-test/semver` when you add/remove item(s)

But perhaps I could add a new `libc-test/semver/emscripten.txt` file in that PR?

**Comment by RReverser** [comment]
> The libc PR was marked as draft because it does not meet this check:
> 
> >

I don't think that's necessary a criteria for draft, it just will be part of review comments? Idk, I just almost never use the draft PR feature so was curious about your usage / rationale.

**Comment by kleisauke** [comment]
> I don't think that's necessary a criteria for draft, it just will be part of review comments?

Ah, sorry I misread your "review straight away" as "merge straight away".

I usually open pull requests as a draft if I think it's not ready to merge yet, these PRs can be reviewed but not merged as-is (because it is dependent on another PR or still being worked on).

Perhaps I should treat draft PRs as non-reviewable rather than non-mergeable? If so, the draft status of these PRs can be removed.

**Comment by RReverser** [comment]
> should treat draft PRs as non-reviewable

Ah that probably is our mismatch - that's how I understood them since they were introduced, yeah, sort of "work in progress, please don't review yet". After all, the button on the draft PRs is called "Ready for review".

**Comment by kleisauke** [comment]
> I just opened draft PR https://github.com/rust-lang/rust/pull/107221 and https://github.com/rust-lang/libc/pull/3087 for this.

It seems that a release of libc with that PR included is taking longer than expected. I'll merge this as-is after the CI passes (the Rust nightly version is pinned, so I don't expect any build failures).

**Review by kleisauke — APPROVED** [review]
Thank you! This will be in v0.0.5.

**Pull request merged**
